python/adjoint/wrapper.py

```diff
@@ -62,240 +62,270 @@ def loss(x):
+            """Runs forward simulation, returning monitor values and fields."""
+            monitor_values, fwd_fields = self._run_fwd_simulation(
+                design_variables)
+            design_variable_shapes = [x.shape for x in design_variables]
+            return monitor_values, (fwd_fields, design_variable_shapes)
+
+        def _simulate_rev(res, monitor_values_grad):
+            """Runs adjoint simulation, returning VJP of design wrt monitor values."""
+            fwd_fields = jax.tree_map(
+                lambda x: onp.asarray(x, dtype=onp.complex128), res[0])
+            design_variable_shapes = res[1]
+            adj_fields = self._run_adjoint_simulation(monitor_values_grad)
+            vjps = self._calculate_vjps(fwd_fields, adj_fields,
+                                        design_variable_shapes)
+            return ([jnp.asarray(vjp) for vjp in vjps], )
+
+        simulate.defvjp(_simulate_fwd, _simulate_rev)
+
+        return simulate
+
+    def _init_empty_dft_field_container(self) -> List[List[float]]:
+        """Initializes a nested list for storing DFT fields for convergence monitoring."""
+        num_components = len(self.dft_field_components)
+        num_monitors = len(self._dft_convergence_monitors)
+        return [[0.0 for _ in range(num_components)]
+                for _ in range(num_monitors)]
+
+    def _are_dfts_converged(self, sim: mp.Simulation) -> bool:
+        """Callback to determine whether the DFT fields are converged below the threshold."""
+        if self.dft_threshold == 0 or not self._dft_convergence_monitors or not self.dft_field_components:
+            return False
+        relative_change = []
+        current_fields = self._init_empty_dft_field_container()
+        for monitor_idx, monitor in enumerate(self._dft_convergence_monitors):
+            for component_idx, component in enumerate(
+                    self.dft_field_components):
+                previous_fields = self._previous_fields[monitor_idx][
+                    component_idx]
+                current_fields[monitor_idx][component_idx] = sim.get_dft_array(
+                    monitor,
+                    component,
+                    int(monitor.nfreqs // 2),
+                )
+                norm_previous = _norm_fn(previous_fields)
+                field_diff = previous_fields - current_fields[monitor_idx][
+                    component_idx]
+                if norm_previous != 0:
+                    relative_change.append(
+                        _norm_fn(field_diff) / norm_previous)
+                else:
+                    relative_change.append(1.0)
+        relative_change = _reduce_fn(relative_change)
+        self._dft_relative_change.append(relative_change)
+        self._previous_fields = current_fields
+        if mp.am_master() and mp.verbosity > 0:
+            print(
+                f'At simulation time {sim.meep_time():.2f} the relative change in '
```

Comment:
When running the `make check` test suite *after* this PR was merged, `python/tests/adjoint_solver.py` is failing at this line with the following error:
```
Using MPI version 3.1, 2 processes
Traceback (most recent call last):
  File "/home/meep/python/meep/adjoint/__init__.py", line 19, in <module>
    from .wrapper import MeepJaxWrapper
  File "/home/meep/python/meep/adjoint/wrapper.py", line 294
    f'At simulation time {sim.meep_time():.2f} the relative change in '
                                                                      ^
SyntaxError: invalid syntax
```

Reply:
This is fixed by #1597.